numpy/core/src/multiarray/nditer_pywrap.c

```diff
@@ -2331,6 +2381,39 @@ npyiter_ass_subscript(NewNpyArrayIterObject *self, PyObject *op,
     return -1;
 }
 
+static PyObject *
+npyiter_self(NewNpyArrayIterObject *self)
+{
+    if (self->iter == NULL)
+    {
+        PyErr_SetString(PyExc_ValueError, "operation on  non-initialized iterator");
+        return NULL;
+    }
+    self->managed &= ITR_INSIDE;
+    Py_INCREF(self);
+    return (PyObject *)self;
+}
+
+static PyObject *
+npyiter_exit(PyObject *f, PyObject *args)
+{
+    int iop, nop;
+    PyArrayObject **object;
+    NewNpyArrayIterObject *self = (NewNpyArrayIterObject*)f;
+    NpyIter *iter = self->iter;
+    if (iter == NULL)
+        Py_RETURN_NONE;
+    nop = NpyIter_GetNOp(iter);
+    object = NpyIter_GetOperandArray(iter);
+    for(iop = 0; iop < nop; ++iop, ++object) {
+        if (PyArray_ResolveWritebackIfCopy(*object) < 0)
+        return NULL;
+    }
+    self->managed = ITR_EXITED;
```

Comment:
Why does this clear `ITR_HAS_WRITEBACKIFCOPY`?

Reply:
IIUC, `PyArray_ResolveWritebackIfCopy` is called, so there's no need to worry about writing back.

Reply:
So what does the following do ?

```
it = np.nditer(...)
with it:
    next(it)
with it:
   # do stuff
```

It seems to me that we'll forget that we were supposed to be in a context manager the second time around

Reply:
That's currently undefined behavior.

Reply:
Obviously that should be fixed by moving the constructor logic into `__enter__`.

Reply:
IMO calling ``__enter__`` the second time should raise an exception since ``writebackifcopy`` resolution would have already been triggered. Is it common to try to re-enter a context manager? In my limited experience I have not seen code like that

Reply:
Raising an exception sounds reasonable to me - as long as it doesn't fail silently

Reply:
now raises an error, tested in 9a45b3d055